core/services/relay/evm/mercury/v3/data_source.go

```diff
@@ -0,0 +1,231 @@
+package mercury_v3
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/big"
+	"sync"
+
+	pkgerrors "github.com/pkg/errors"
+
+	ocrtypes "github.com/smartcontractkit/libocr/offchainreporting2plus/types"
+
+	relaymercury "github.com/smartcontractkit/chainlink-relay/pkg/reportingplugins/mercury"
+	relaymercuryv3 "github.com/smartcontractkit/chainlink-relay/pkg/reportingplugins/mercury/v3"
+
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/services/job"
+	"github.com/smartcontractkit/chainlink/v2/core/services/ocrcommon"
+	"github.com/smartcontractkit/chainlink/v2/core/services/pipeline"
+	"github.com/smartcontractkit/chainlink/v2/core/services/relay/evm/mercury/types"
+	"github.com/smartcontractkit/chainlink/v2/core/utils"
+)
+
+type Runner interface {
+	ExecuteRun(ctx context.Context, spec pipeline.Spec, vars pipeline.Vars, l logger.Logger) (run pipeline.Run, trrs pipeline.TaskRunResults, err error)
+}
+
+type LatestReportFetcher interface {
+	LatestPrice(ctx context.Context, feedID [32]byte) (*big.Int, error)
+	LatestTimestamp(context.Context) (uint32, error)
+}
+
+type datasource struct {
+	pipelineRunner Runner
+	jb             job.Job
+	spec           pipeline.Spec
+	feedID         types.FeedID
+	lggr           logger.Logger
+	runResults     chan<- pipeline.Run
+
+	fetcher      LatestReportFetcher
+	linkFeedID   types.FeedID
+	nativeFeedID types.FeedID
+
+	mu sync.RWMutex
+
+	chEnhancedTelem chan<- ocrcommon.EnhancedTelemetryMercuryData
+}
+
+var _ relaymercuryv3.DataSource = &datasource{}
+
+var maxInt192 = new(big.Int).Exp(big.NewInt(2), big.NewInt(191), nil)
+
+func NewDataSource(pr pipeline.Runner, jb job.Job, spec pipeline.Spec, feedID types.FeedID, lggr logger.Logger, rr chan pipeline.Run, enhancedTelemChan chan ocrcommon.EnhancedTelemetryMercuryData, fetcher LatestReportFetcher, linkFeedID, nativeFeedID types.FeedID) *datasource {
+	return &datasource{pr, jb, spec, feedID, lggr, rr, fetcher, linkFeedID, nativeFeedID, sync.RWMutex{}, enhancedTelemChan}
+}
+
+func (ds *datasource) Observe(ctx context.Context, repts ocrtypes.ReportTimestamp, fetchMaxFinalizedTimestamp bool) (obs relaymercuryv3.Observation, merr error) {
+	run, trrs, err := ds.executeRun(ctx)
```

Comment:
Is this not executed concurrently within the waitgroup?

Reply:
The check down below allows to skip fetching prices from the server if the the `linkFeedId` or `nativeFeedId` equal to the feedId. So we needed a price to be known at this point.